### ISSUE TITLE
How to unregister from FCM (rather than unsubscribe from a topic)

### ISSUE THREAD

**Issue by idekterev** [fechanique/cordova-plugin-fcm#353]
I have a situation where I would like to stop receiving push notifications on certain events (user log-out). How can let FCM know that I do not want to receive push notifications via this plugin (de-register from FCM)?

I have not subscribed to any topics.

**Comment by blumanski** [comment]
How do you send  the notifications? I handle those thinks all in my backend api.

**Comment by TamkeenLMS** [comment]
Yes, I am wondering about this too. You don't want the user to receive push notification after he has logged out!

**Comment by idekterev** [comment]
I'm sending the push notification using the registration_ids parameter (details from here: https://firebase.google.com/docs/cloud-messaging/http-server-ref)

**Comment by blumanski** [comment]
I see the whole thing from a different perspective. 

The registration_id (device token), is only to allow your device to receive notifications.
In your backend api, you can implement what ever you want to send messages to what device.

I use a db table with "channels" and devices are assigned to a channel or not, if you use it that way, you can just get all devices assigned to the channel and send a notification to all of the devices.
And with that, you can have a flag in your db like "on/off" to know if a device should get a notification or not.

So, if user logs out, api call to logout and set flag to "device off" -> if user logs in, set flag to "device on".

Even if you have only one channel, just have a table with one channel and assign the device to it.

This way, you use the device token to send the notifications, which is preferred over the use of "channel subscription".

I am using it this way with unlimited channels.

To send notifications I use this project
https://github.com/Paragraph1/php-fcm

**Comment by idekterev** [comment]
Yes I do understand your point - the sending/not sending could be managed on the server. In my app there are numerous events that cause a 'pause' of receiving notifications. I was trying to find a way of managing this on the client also without introducing a logout/pause http call every time. I take it there is little chance of seeing this feature then?

**Comment by ostownsville** [comment]
Is probably easy to add @fechanique

    FirebaseInstanceId.getInstance().deleteInstanceId();

**Comment by mreis1** [comment]
+1 to unregister the device from all FCM subscribed topics.

**Comment by NodeJoSe** [comment]
@idekterev did you find any solution or workaround for this? 
I tried to delete the storage, but I can't find a way to remove the entire storage from the device.

I'm using ionic-storage, and using `Storage.clear()` does not remove the entire app data, I can only delete the instanceId only by removing the app data manually in the device "manage applications" option, is there a way to remove the entire app storage from the device using ionic?

Also tried ` window.localStorage.clear();` and `window.sessionStorage.clear();` but neither of those works.

**Comment by idekterev** [comment]
@NodeJoSe - sorry no proper solution as yet. I changed some logic on the back end as a workaround. P.s.the FCM instanceId is not stored in localStorage, so no javascript functions will affect it. (Unless the JS instructs the cordova fcm plugin to do this action - which was my initial suggestion).

In fact @fechanique, I have another use case: I would like to be able to re-subscribe to FCM with a different ID (e.g. every time I log in)- is this possible?

**Comment by dagalti** [comment]
@idekterev  delete subscription from fcm database. 

To delete a push subscription, send a DELETE request to:

` https://iid.googleapis.com/v1/web/iid/REGISTRATION_TOKEN`

Example DELETE request
` https://iid.googleapis.com/v1/web/iid/KctODamlM4:CI2k_HHw...3P1
 Authorization:key=AIzaSyZ-1u...0GBYzPu7Udno5aA`

Example result
` HTTP 204 No-Content`

PHP CURL Example

 function delete_subscription ($regtoken){
 $url = 'https://iid.googleapis.com/v1/web/iid/' . $regtoken;
  $headers = array ('Authorization: key=' . 'AUTH_KEY');
  $ch = curl_init ();
  curl_setopt ($ch, CURLOPT_URL, $url );
  curl_setopt ($ch, CURLOPT_CUSTOMREQUEST, "DELETE");
  curl_setopt ($ch, CURLOPT_HTTPHEADER, $headers );
  curl_setopt ($ch, CURLOPT_RETURNTRANSFER, true );
   $result = curl_exec ( $ch );
  curl_close ( $ch );
   return $result;
  }
 
Check this link : https://developers.google.com/instance-id/reference/server#delete_push_subscriptions

**Comment by idekterev** [comment]
@dagalti - thank you. Very helpful.

@NodeJoSe - I switched to using this plugin (https://github.com/phonegap/phonegap-plugin-push) instead, which allows the client to unsubscribe, and some other options.

**Comment by sam43** [comment]
You can do it with kotlin coroutines as well using thread : 
    GlobalScope.launch {
        FirebaseInstanceId.getInstance().deleteInstanceId()
    }

**Comment by hlmifzi** [comment]
very helpful @dagalti , i confused, if i get token but is not registered

**Comment by himeshagile** [comment]
> @idekterev delete subscription from fcm database.
> 
> To delete a push subscription, send a DELETE request to:
> 
> ` https://iid.googleapis.com/v1/web/iid/REGISTRATION_TOKEN`
> 
> Example DELETE request
> ` https://iid.googleapis.com/v1/web/iid/KctODamlM4:CI2k_HHw...3P1 Authorization:key=AIzaSyZ-1u...0GBYzPu7Udno5aA`
> 
> Example result
> ` HTTP 204 No-Content`
> 
> PHP CURL Example
> 
> function delete_subscription ($regtoken){
> $url = 'https://iid.googleapis.com/v1/web/iid/' . $regtoken;
> $headers = array ('Authorization: key=' . 'AUTH_KEY');
> $ch = curl_init ();
> curl_setopt ($ch, CURLOPT_URL, $url );
> curl_setopt ($ch, CURLOPT_CUSTOMREQUEST, "DELETE");
> curl_setopt ($ch, CURLOPT_HTTPHEADER, $headers );
> curl_setopt ($ch, CURLOPT_RETURNTRANSFER, true );
> $result = curl_exec ( $ch );
> curl_close ( $ch );
> return $result;
> }
> 
> Check this link : https://developers.google.com/instance-id/reference/server#delete_push_subscriptions
id/reference/server#delete_push_subscriptions

Request is missing required authentication credential. Expected OAuth 2 access token rathere then serverkey

**Comment by ZeeshanAhmadKhalil** [comment]
> Is probably easy to add @fechanique
> 
> ```
> FirebaseInstanceId.getInstance().deleteInstanceId();
> ```

There are some problems in that case;

1.  If that token has been subbed to topics, how I will re-sub to those topics after user logins again?
2. What if the same user logins with a different device, in that case, his reg token will also change.
